### ISSUE TITLE
Significant memory overhead in `write_parquet`

### ISSUE THREAD

**Issue by sagiahrac** [Eventual-Inc/Daft#3443]
### Describe the bug

When generating a 1 GiB dataset, the peak memory usage varies significantly depending on the daft data types (`uint8`, `list[uint8]`, `tensor[uint8]`). The memory spike occurs only during `write_parquet`.

#### Steps:
1. Create a dataset of size 1 GiB (using 1024**3 `uint8` elements).
2. Store the data in three formats:
   - `uint8`
   - `list[uint8]`
   - `tensor[uint8]`
3. Write each dataset to disk using daft’s `write_parquet` method.

#### Code:
```python
import tempfile
import numpy as np
import daft
from typing import Literal


def gen_1gb_df(dtype: Literal["uint8", "list[uint8]", "tensor[uint8]"]) -> daft.DataFrame:
    if dtype == "uint8":
        shape = (1024 * 1024 * 1024)
    elif dtype == "list[uint8]":
        shape = (1024, 1024 * 1024)
    elif dtype == "tensor[uint8]":
        shape = (1024, 1024, 1024)
    else:
        raise ValueError(f"Invalid dtype: {dtype}")
        
    df = daft.from_pydict({"A": np.random.randint(low=0, high=256, size=shape, dtype=np.uint8)})  # 1GB
    return df

def write_temp_parquet(df: daft.DataFrame):
    with tempfile.TemporaryDirectory() as tmpdir:
        df.write_parquet(tmpdir)


if __name__ == "__main__":
    # daft.set_execution_config(enable_native_executor=True)
    df = gen_1gb_df(DTYPE)
    write_temp_parquet(df)
```

#### Results:
`uint8`:
![image](https://github.com/user-attachments/assets/05ef43f5-3497-45bb-822b-523413fa626e)

`list[uint8]`:
![image](https://github.com/user-attachments/assets/17b70a0e-53a6-43bb-870d-c2b9e3f61b71)

`tensor[uint8]`:
![image](https://github.com/user-attachments/assets/cc0b2469-db64-44bf-99d4-5d25402024ab)

#### Reuslts with native executor:
If adding the following line before generating the df
```python
daft.set_execution_config(enable_native_executor=True)
```
`uint8`:
![image](https://github.com/user-attachments/assets/5b229892-835b-4fa7-b3fa-45eb87320c6b)

`list[uint8]`:
![image](https://github.com/user-attachments/assets/63642d65-5a26-4cb2-b093-37bbfa553337)

`tensor[uint8]`:
![image](https://github.com/user-attachments/assets/3cba6d21-997a-4002-adac-db51c1231aec)

### Expected behavior

A x20 or x35 memory overhead (depends on the runner) for certain datatypes seems excessive.

### Component(s)

Parquet

### Additional context

Materializing the dataframe without writing it results in much lower memory usage.
I was using `daft v0.3.14`

**Comment by colin-ho** [comment]
Thanks for making this issue!

I've ran the repro and I've identified a couple issues:
1. The write operator in swordfish currently buffers data into `morsel_size` rows and subsequently into `rows_per_row_group` rows. This double buffering is unnecessary and we just need to do `rows_per_row_group` rows. 
2. Instead of buffering by `rows_per_row_group` rows we should buffer by a strict `row_group_size` bytes. In the list/tensor examples there are only 1024 rows, but each row is very large. The py_runner is able to write this into 3 row groups, but swordfish only writes 1 row group. This causes inflated memory usage because swordfish needs to allocate a much larger buffer to write this 1 mega row group.

Should have a fix for this pretty soon.

This is a before and after of the `tensor[uint8]` example with the above fixes implemented. (Running swordfish on a 128cpu machine)

<img width="1194" alt="Screenshot 2024-11-27 at 10 26 07 AM" src="https://github.com/user-attachments/assets/668a065a-1748-41dd-b511-c8f5e9721d5f">

<img width="1195" alt="Screenshot 2024-11-27 at 10 12 27 AM" src="https://github.com/user-attachments/assets/8297ae9d-b5af-43bd-8b7f-b75655158e15">

**Comment by sagiahrac** [comment]
That was quick - great catch! Thanks!

**Comment by kevinzwang** [comment]
@sagiahrac looks like a fix for this issue has been implemented and released in version 0.3.16. I'll close this issue now, feel free to reopen if the issue persists in the new library version.

**Comment by kevinzwang** [comment]
Sorry what I meant to say is that the change is in main and _will_ be part of the upcoming v0.3.16 release. We have not yet published v0.3.16.
